sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs

```diff
@@ -57,42 +58,84 @@ protected override void Initialize(SonarAnalysisContext context)
 
                     var literal = (LiteralExpressionSyntax)c.Node;
 
-                    var numberWithoutSuffix = ClearNumberTypeSuffix(literal.Token.Text);
-
-                    var decimalParts = numberWithoutSuffix.Split('.');
-                    if (decimalParts.Length > 2)
-                    {
-                        return;
-                    }
-
-                    var hasIrregularPattern = decimalParts
-                        .SelectMany(part => part.Split('_'))
-                        .Select(x => x.Length)
-                        .Skip(1) // skip the first part (1_234 => 234)
-                        .Reverse().Skip(decimalParts.Length == 2 ? 1 : 0) // skip the last if there is a decimal (.234_5 => 234)
-                        .Distinct().Skip(1).Any(); // we expect to have only 1 size of pattern
-
-                    if (hasIrregularPattern)
+                    if (HasIrregularPattern(literal.Token.Text))
                     {
                         c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, literal.GetLocation()));
                     }
                 },
                 SyntaxKind.NumericLiteralExpression);
         }
 
-        private static string ClearNumberTypeSuffix(string numberLiteral)
+        internal static bool HasIrregularPattern(string numericToken)
         {
-            if (numberLiteral.EndsWith("UL", StringComparison.OrdinalIgnoreCase))
+            var splitted = StripNumericPreAndSuffix(numericToken).Split(Dot);
```

Comment:
please use `split` (the participle and past tense forms are the same as the present form)